source/decode.c

```diff
@@ -404,48 +467,89 @@ static int s_linestate_header(struct aws_http_decoder *decoder, struct aws_byte_
             }
-            for (size_t i = 0; i < n; ++i) {
-                struct aws_byte_cursor coding = s_trim_whitespace(codings[i]);
+
+            /* RFC-7230 section 3.3.1 Transfer-Encoding */
+            /* RFC-7230 section 4.2 Compression Codings */
+
+            /* Note that it's possible for multiple Transfer-Encoding headers to exist, in which case the values
+             * should be appended with those from any previously encountered Transfer-Encoding headers. */
+            struct aws_byte_cursor split;
+            AWS_ZERO_STRUCT(split);
+            while (aws_byte_cursor_next_split(&header.value_data, ',', &split)) {
+                struct aws_byte_cursor coding = s_trim_whitespace(split);
+                int prev_flags = decoder->transfer_encoding;
+
                 if (aws_string_eq_byte_cursor_ignore_case(s_transfer_coding_chunked, &coding)) {
-                    flags |= AWS_HTTP_TRANSFER_ENCODING_CHUNKED;
-                    if (i != (n - 1)) {
-                        /* chunked must be the final transfer coding */
-                        return aws_raise_error(AWS_ERROR_HTTP_PARSE);
-                    }
-                } else if (aws_string_eq_byte_cursor_ignore_case(s_transfer_coding_compress, &coding)) {
-                    flags |= AWS_HTTP_TRANSFER_ENCODING_DEPRECATED_COMPRESS;
+                    decoder->transfer_encoding |= AWS_HTTP_TRANSFER_ENCODING_CHUNKED;
+
+                } else if (
+                    aws_string_eq_byte_cursor_ignore_case(s_transfer_coding_compress, &coding) ||
+                    aws_string_eq_byte_cursor_ignore_case(s_transfer_coding_x_compress, &coding)) {
+                    /* A recipient SHOULD consider "x-compress" to be equivalent to "compress". RFC-7230 4.2.1 */
+                    decoder->transfer_encoding |= AWS_HTTP_TRANSFER_ENCODING_DEPRECATED_COMPRESS;
+
                 } else if (aws_string_eq_byte_cursor_ignore_case(s_transfer_coding_deflate, &coding)) {
-                    flags |= AWS_HTTP_TRANSFER_ENCODING_DEFLATE;
-                } else if (aws_string_eq_byte_cursor_ignore_case(s_transfer_coding_gzip, &coding)) {
-                    flags |= AWS_HTTP_TRANSFER_ENCODING_GZIP;
+                    decoder->transfer_encoding |= AWS_HTTP_TRANSFER_ENCODING_DEFLATE;
+
+                } else if (
+                    aws_string_eq_byte_cursor_ignore_case(s_transfer_coding_gzip, &coding) ||
+                    aws_string_eq_byte_cursor_ignore_case(s_transfer_coding_x_gzip, &coding)) {
+                    /* A recipient SHOULD consider "x-gzip" to be equivalent to "gzip". RFC-7230 4.2.3 */
+                    decoder->transfer_encoding |= AWS_HTTP_TRANSFER_ENCODING_GZIP;
+
+                } else if (coding.len > 0) {
+                    AWS_LOGF_ERROR(
+                        AWS_LS_HTTP_STREAM,
+                        "id=%p: Incoming transfer-encoding header lists unrecognized coding.",
```

Comment:
[nitpick] Wouldn't it be an unrecognied encoding?

Reply:
They come in a comma separated list, and apparently empty entries are ok

Reply:
I just mean linguistically :)